internal/charmstore/migrations_test.go

```diff
@@ -825,49 +333,55 @@ func (s *migrationsSuite) checkBaseEntitiesCount(c *gc.C, expectCount int) {
 	c.Assert(count, gc.Equals, expectCount)
 }
 
-func (s *migrationsSuite) insertEntity(c *gc.C, id *charm.Reference, name string, size int64) {
-	entity := &mongodoc.Entity{
-		URL:     id,
-		BaseURL: baseURL(id),
-		Name:    name,
-		Size:    size,
+// denormalizeEntity returns a copy of e0 with all denormalized fields
+// filled out if they are zero.
+func denormalizeEntity(e0 *mongodoc.Entity) *mongodoc.Entity {
+	e := *e0
+	if e.BaseURL == nil {
+		e.BaseURL = baseURL(e.URL)
 	}
-	err := s.db.Entities().Insert(entity)
-	c.Assert(err, gc.IsNil)
-
-	// Remove the denormalized fields if required.
-	if name != "" {
-		return
+	if e.Name == "" {
+		e.Name = e.URL.Name
 	}
-	err = s.db.Entities().UpdateId(id, bson.D{{
-		"$unset", bson.D{
-			{"user", true},
-			{"name", true},
-			{"revision", true},
-			{"series", true},
-		},
-	}})
-	c.Assert(err, gc.IsNil)
-}
-
-func (s *migrationsSuite) insertBaseEntity(c *gc.C, id *charm.Reference, acls *mongodoc.ACL) {
-	entity := &mongodoc.BaseEntity{
-		URL:    id,
-		Name:   id.Name,
-		User:   id.User,
-		Public: true,
+	if e.User == "" {
```

Comment:
It will happen whenever someone passes an entity with an empty User field (which is the common case)